### PR TITLE
Upgrade babel to 7.x.x and jest to 27.x.x

### PR DESCRIPTION
**Pull request by daltojohnso**
This PR upgrades Babel and Jest to their most recent releases.

I recommend going commit by commit to understand what all needed changing. I tried to keep the commits clean and informative.

I couldn't figure out how to make Babel 7 work with our ancient version of Jest. Upgrading Jest doesn't seem like an additional risk, so I figured spending time making the latest version of Jest work would be a better investment.

This change does come with some risk since it changes how our JS is transformed. We will need to test the app extensively since functions that only run rarely may now contain code that is broken.

We could potentially reduce _some_ risk by adding more support for older browsers to handle situations where our code doesn't match the current spec. For example, https://github.com/metabase/metabase/pull/17137/commits/b238828831f110f6af5ebfd47c65065f8e830cfe, which was found due to a broken e2e test, wouldn't break if we transformed our `const` declarations into `var` declarations (what we do right now), since `var` declarations are hoisted. I'm on the fence with regards to doing that, so happy to hear what others think.

### CODE REVIEW

File: frontend/test/metabase-lib/lib/Dimension.unit.spec.js [hunk 1]
```diff
@@ -138,26 +138,6 @@ describe("Dimension", () => {
...
   describe("INSTANCE METHODS", () => {
-    describe("dimensions()", () => {
-      it("returns `dimension_options` of the underlying field if available", () => {
-        pending();
-      });
-      it("returns sub-dimensions for matching dimension if no `dimension_options`", () => {
-        // just a single scenario should be sufficient here as we will test
-        // `static dimensions()` individually for each dimension
-        pending();
-      });
```
Comment (kulyk) [inline frontend/test/metabase-lib/lib/Dimension.unit.spec.js:149]: Optionally, we can use [`test.todo("...")`](https://jestjs.io/docs/api#testtodoname) if we want to keep the tests

**Review by kulyk — APPROVED** [review]
Code looks good, made a few smoke tests within the branch, and couldn't spot any regressins. I'm the most worried about [`MBQLClause` `map` and `filter` returning plain `Arrays` instead of `MBQLClause`](https://github.com/metabase/metabase/pull/17137/commits/b7dd7f67a0d2aeffee64d015c69c7544552889c1). However I couldn't easily find occurrences of that, but they could be hidden somewhere

Great work @daltojohnso 🥇 Thank you!

**Comment by nemanjaglumac** [comment]
When doing `yarn dev-ee`, I'm receiving the following error for docs.

<details>

```
[docs] (node:66597) Warning: Accessing non-existent property 'cat' of module exports inside circular dependency
[docs] (Use `node --trace-warnings ...` to show where the warning was created)
[docs] (node:66597) Warning: Accessing non-existent property 'cd' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'chmod' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'cp' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'dirs' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'pushd' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'popd' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'echo' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'tempdir' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'pwd' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'exec' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'ls' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'find' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'grep' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'head' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'ln' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'mkdir' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'rm' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'mv' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'sed' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'set' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'sort' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'tail' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'test' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'to' of module exports inside circular dependency
$ husky install
[docs] (node:66597) Warning: Accessing non-existent property 'toEnd' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'touch' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'uniq' of module exports inside circular dependency
[docs] (node:66597) Warning: Accessing non-existent property 'which' of module exports inside circular dependency
[frontend] husky - Git hooks installed
[docs] Error: Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel. while parsing file: /path/to/metabase/frontend/src/metabase-lib/lib/Action.js
[docs]     at throwVersionError (/path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:78:11)
[docs]     at Object.assertVersion (/path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:28:5)
[docs]     at /path/to/metabase/node_modules/@babel/plugin-transform-flow-strip-types/lib/index.js:15:7
[docs]     at /path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:22:12
[docs]     at Function.memoisePluginContainer (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:113:13)
[docs]     at Function.normalisePlugin (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:146:32)
[docs]     at /path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:184:30
[docs]     at Array.map (<anonymous>)
[docs]     at Function.normalisePlugins (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
[docs]     at OptionManager.mergeOptions (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
[docs]     at OptionManager.init (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
[docs]     at File.initOptions (/path/to/metabase/node_modules/babel-core/lib/transformation/file/index.js:212:65)
[docs]     at new File (/path/to/metabase/node_modules/babel-core/lib/transformation/file/index.js:135:24)
[docs]     at Pipeline.transform (/path/to/metabase/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
[docs]     at Babelify._flush (/path/to/metabase/node_modules/babelify/index.js:27:24)
[docs]     at Babelify.prefinish (internal/streams/transform.js:147:10)
[docs]     at Babelify.emit (events.js:375:28)
[docs]     at prefinish (internal/streams/writable.js:630:14)
[docs]     at finishMaybe (internal/streams/writable.js:638:5)
[docs]     at Babelify.Writable.end (internal/streams/writable.js:582:5)
[docs]     at DestroyableTransform.onend (/path/to/metabase/node_modules/readable-stream/lib/_stream_readable.js:577:10)
[docs]     at Object.onceWrapper (events.js:481:28)
[docs]     at DestroyableTransform.emit (events.js:387:35)
[docs]     at endReadableNT (/path/to/metabase/node_modules/readable-stream/lib/_stream_readable.js:1010:12)
[docs]     at processTicksAndRejections (internal/process/task_queues.js:82:21) {
[docs]   code: 'BABEL_VERSION_UNSUPPORTED',
[docs]   version: '6.26.3',
[docs]   range: '^7.0.0-0',
[docs]   filename: '/path/to/metabase/frontend/src/metabase-lib/lib/Action.js'
[docs] }
[docs] documentation.js serving on port 4001
[docs] events.js:352
[docs]       throw er; // Unhandled 'error' event
[docs]       ^
[docs]
[docs] Error: Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel. while parsing file: /path/to/metabase/frontend/src/metabase-lib/lib/Dashboard.js
[docs]     at throwVersionError (/path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:78:11)
[docs]     at Object.assertVersion (/path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:28:5)
[docs]     at /path/to/metabase/node_modules/@babel/plugin-transform-flow-strip-types/lib/index.js:15:7
[docs]     at /path/to/metabase/node_modules/@babel/helper-plugin-utils/lib/index.js:22:12
[docs]     at Function.memoisePluginContainer (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:113:13)
[docs]     at Function.normalisePlugin (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:146:32)
[docs]     at /path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:184:30
[docs]     at Array.map (<anonymous>)
[docs]     at Function.normalisePlugins (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
[docs]     at OptionManager.mergeOptions (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
[docs]     at OptionManager.init (/path/to/metabase/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
[docs]     at File.initOptions (/path/to/metabase/node_modules/babel-core/lib/transformation/file/index.js:212:65)
[docs]     at new File (/path/to/metabase/node_modules/babel-core/lib/transformation/file/index.js:135:24)
[docs]     at Pipeline.transform (/path/to/metabase/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
[docs]     at Babelify._flush (/path/to/metabase/node_modules/babelify/index.js:27:24)
[docs]     at Babelify.prefinish (internal/streams/transform.js:147:10)
[docs]     at Babelify.emit (events.js:375:28)
[docs]     at prefinish (internal/streams/writable.js:630:14)
[docs]     at finishMaybe (internal/streams/writable.js:638:5)
[docs]     at Babelify.Writable.end (internal/streams/writable.js:582:5)
[docs]     at DestroyableTransform.onend (/path/to/metabase/node_modules/readable-stream/lib/_stream_readable.js:577:10)
[docs]     at Object.onceWrapper (events.js:481:28)
[docs]     at DestroyableTransform.emit (events.js:387:35)
[docs]     at endReadableNT (/path/to/metabase/node_modules/readable-stream/lib/_stream_readable.js:1010:12)
[docs]     at processTicksAndRejections (internal/process/task_queues.js:82:21)
[docs] Emitted 'error' event on Deps instance at:
[docs]     at Babelify.<anonymous> (/path/to/metabase/node_modules/module-deps-sortable/index.js:259:18)
[docs]     at Babelify.emit (events.js:387:35)
[docs]     at Babelify._flush (/path/to/metabase/node_modules/babelify/index.js:32:10)
[docs]     at Babelify.prefinish (internal/streams/transform.js:147:10)
[docs]     [... lines matching original stack trace ...]
[docs]     at Object.onceWrapper (events.js:481:28) {
[docs]   code: 'BABEL_VERSION_UNSUPPORTED',
[docs]   version: '6.26.3',
[docs]   range: '^7.0.0-0',
[docs]   filename: '/path/to/metabase/frontend/src/metabase-lib/lib/Dashboard.js'
[docs] }
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>

**Comment by daltojohnso** [comment]
@kulyk -- I agree that the MBQLArrayClause change is one of the riskier ones. There is a `loose` config option for `plugin-transform-classes` (see https://babeljs.io/docs/en/babel-plugin-transform-classes) that may transform this code into something closer to what it is currently with Babel 6, so I will test it out today.

File: frontend/test/metabase-lib/lib/Dimension.unit.spec.js (hunk 1 above)
Comment (daltojohnso) [reply frontend/test/metabase-lib/lib/Dimension.unit.spec.js:149]: Yeah, I lean toward blindly removing code that has been dead for 3+ years. I'm not sure how many of these test descriptions are outdates/no longer relevant. Whoever ends up writing all the unit tests for this code (hopefully soon 🙂 ) should probably approach it with fresh eyes.

**Comment by daltojohnso** [comment]
This change -- https://github.com/metabase/metabase/pull/17137/commits/fa6860f730557d0274e98353595294871a797226 -- maintains current `MBQLArrayClause` behavior.

Interestingly, removing the browserslist (so that Babel transpiles to ES2015) or removing `"not IE 11"` from the browserslist causes one unit test to break due to an error in a function named `_objectWithoutPropertiesLoose` that is used in destructuring things like `{foo, ...rest}`. Specifically, the line it is breaking on is:

```
entity.Form = ({ object, ...props }) => (
```

in `metabase/entities/containers/index.js`.

Not sure why it is breaking.

**Comment by daltojohnso** [comment]
Setting `browserslist` to `["defaults"]` means our build will support features for the following browsers:

```
and_chr 90
and_ff 87
and_qq 10.4
and_uc 12.12
android 90
baidu 7.12
chrome 90
chrome 89
chrome 88
edge 90
edge 89
firefox 88
firefox 87
firefox 78
ie 11
ios_saf 14.5
ios_saf 14.0-14.4
ios_saf 13.4-13.7
kaios 2.5
op_mini all
op_mob 62
opera 75
opera 74
safari 14.1
safari 14
samsung 14.0
samsung 13.0
```

You can generate this list by running `npx browserslist`

**Pull request merged**